### DISCUSSION TITLE
Deployment guide: AWS

### DISCUSSION THREAD

**Discussion by michaelbromley** [vendurehq/vendure#1534] [Deployment]
Collecting information / tips / gotchas about deploying Vendure to AWS. The idea is to consolidate into an official guide.

**Comment by flowfalls** [comment]
I'll write more later on but with my use case I utilize the following:

- Docker backend (worker and core vendure) using https://docs.docker.com/cloud/ecs-integration/ (Fargate ECS)
- I have a lambda and cloudwatch event to stop the worker during late hours (cost reduction)
- I use RDS
- I use Opensearch (formerly ElasticSearch) - this is the most expensive component so be wary if a startup
- I use Route53, Cloudfront + S3 with SPA

I have gone down the route of manually pushing with docker ecs-integration but may reconsider later down the line. 

I'm also interested in using serverless architecture (lambdas for frontend, and possibly backend?) but want to know the performance implications.

Cheers

**Comment by khawarizmus** [comment]
I have not deployed Vuendure to AWS but a cool thing to have for AWS is a [terraform](https://www.terraform.io/) project template with all the needed components that @cribmove mentioned and plugin [infracost](https://www.infracost.io/) to give a cost estimate with a cost policy that is set to fail. that way anyone start a project from that template and just inject their AWs credentials to deploy their Vendure instance and it would give them a cost estimate before deployment.

**Reply by flowfalls** [reply]
I'll try and implement this @gimyboya in Terraform. Without Elasticsearch it defaults to around $15 a month, with Elasticsearch - $30-40. The Infracost looks extremely interesting as well thanks for making me aware of it. 

Thanks

**Comment by casabian** [comment]
My two cents on AWS deployment:

- Deploy server as serverless app (e.g. https://docs.nestjs.com/faq/serverless)
- Deploy  worker as serverless app backed by SQS

I think running all parts of Vendure in a serverless manner could greatly reduce infrastructure cost
 
Btw I'm working on a Vendure plugin that implements JobQueueStrategy for SQS similar to https://github.com/Pinelab-studio/pinelab-vendure-plugins/tree/master/packages/vendure-plugin-google-cloud-tasks)

**Reply by michaelbromley** [reply]
Yes, serverless is definitely interesting from the cost perspective alone. Good to hear you are working on a plugin! I'll be happy to list it on https://www.vendure.io/plugins/ if you decide to open-source it.

If you get somewhere setting up serverless Vendure on AWS, I would love to get your input on a guide.

**Reply by toonjanssens** [reply]
@casabian Did you create a plugin to connect to the AWS SQS?

**Reply by tufail** [reply]
@casabian is there any update on the plugin

**Comment by cermakondrej** [comment]
Has anyone managed to actually deploy the whole stack (Database, Server, Worker, AdminUI, Storefront) to AWS?

My goal was to use NX monorepo with angular universal and vendure, running everything on AWS Lambda to minimize cost and use SQS for worker to minimize costs, however it seems that its quite tricky to setup. 

I've seen [article for setup for Google cloud](https://www.vendure.io/blog/2021/11/serverless-e-commerce-vendure-on-google-cloud-run/), however Google cloud doesn't offer any SQL in free tier which means quite big cost jump from the start (min 40$/month)
